lib/purl/generator.ex

```diff
@@ -0,0 +1,84 @@
+# credo:disable-for-this-file Credo.Check.Refactor.Nesting
+with {:module, StreamData} <- Code.ensure_loaded(StreamData) do
+  defmodule Purl.Generator do
+    @moduledoc """
+    `StreamData` generator for valid purls
+    """
+
+    import StreamData
+
+    alias Purl.Generator.Version, as: VersionGenerator
+
+    @spec type :: StreamData.t(Purl.type())
+    def type, do: require_letter_start(string([?a..?z, ?A..?Z, ?0..?9, ?., ?+, ?-], min_length: 1, max_length: 20))
+
+    @spec namespace_segment :: StreamData.t(Purl.namespace_segment())
+    def namespace_segment, do: filter(string(:printable, min_length: 1, max_length: 50), &(not String.contains?(&1, "/")))
```

Comment:
There are two special cases you are missing here. The namespace segment must not be `.` or `..`. `.` and `..` are illegal namespace segments and should be stripped. It looks like that behavior is implemented in the parser, so the test will fail in the unlikely case that namespace segment is `.` or `..`.

Reply:
Good catch, fixed.

Reply:
(on subpath, not namespace...)